Mingw compilation support + some fixups.

abramhindle opened the pull request:
* fix some case sensitivity in the headers
* jiggle around the CMakeLists.txt to work with mingw without ruining it for normal people
* fix 1 url in the README
* Add mingw details to the readme
* Add mingw cmake stuff to the root (should it be there?)
* Dealt with a bunch of whining about function pointer casts
* Hid SetThreadDescription on MING32
* Converted the dumpfile open to C++11 / C++17 standard (only on windows is wchar open allowed, but there is a call to normalize it)

I did successfully run epic games wonderlands with a mingw compiled OHL dll on Ubuntu with lutris and GE-7.1.

I DID NOT TEST ON WIN32/WIN64. You have to check to make sure I didn't wreck anything for you. I might've.

abramhindle committed:
Adding mingw support to OpenHotfixLoader
* fix some case sensitivity in the headers
* jiggle around the CMakeLists.txt to work with mingw without ruining it
  for normal people
* fix 1 url in the README
* Add mingw details to the readme
* Add mingw cmake stuff to the root (should it be there?)
* Dealt with a bunch of whining about function pointer casts
* Hid SetThreadDescription on MING32
* Converted the dumpfile open to C++11 / C++17 standard (only on windows
  is wchar open allowed, but there is a call to normalize it)

I did successfully run epic games wonderlands with a mingw compiled OHL
dll on Ubuntu with lutris and GE-7.1.

DID NOT TEST ON WIN32/WIN64

abramhindle committed:
mild formatting change

abramhindle committed:
clang format fixes

apple1417 commented:
What version of mingw are you using? Mine's failing to link threads.
```
[ 79%] Linking CXX shared library libOpenHotfixLoader-Debug.dll
/usr/bin/cmake -E cmake_link_script CMakeFiles/OpenHotfixLoader.dir/link.txt --verbose=1
/usr/bin/cmake -E rm -f CMakeFiles/OpenHotfixLoader.dir/objects.a
/usr/bin/x86_64-w64-mingw32-ar qc CMakeFiles/OpenHotfixLoader.dir/objects.a @CMakeFiles/OpenHotfixLoader.dir/objects1.rsp
/usr/bin/x86_64-w64-mingw32-g++-posix -std=c++17  -g -shared -o libOpenHotfixLoader-Debug.dll -Wl,--out-implib,libOpenHotfixLoader-Debug.dll.a -Wl,--major-image-version,0,--minor-image-version,0 -Wl,--whole-archive CMakeFiles/OpenHotfixLoader.dir/objects.a -Wl,--no-whole-archive @CMakeFiles/OpenHotfixLoader.dir/linklibs.rsp
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/OpenHotfixLoader.dir/objects.a(loader.cpp.obj): in function `ohl::loader::reload()':
/mnt/h/OpenHotfixLoader/src/loader.cpp:504: undefined reference to `std::thread::detach()'
```
Could you compile with `set(CMAKE_VERBOSE_MAKEFILE On)` and see what command it's running for you?

FWIW the Windows build still runs fine.

apple1417 commented on mingw-w64-x86_64.cmake:
```diff
@@ -0,0 +1,26 @@
+
+# cross compilers to use for C, C++ and Fortran
+set(CMAKE_C_COMPILER ${TOOLCHAIN_PREFIX}-gcc)
+set(CMAKE_CXX_COMPILER ${TOOLCHAIN_PREFIX}-g++-posix)
```
Any reason this is using `-posix` over the base?

abramhindle replied:
cpr doesn't compile with win32, "base" on ubuntu is -win32. And in OHL std::thread is not defined in win32 mingw.

abramhindle replied:
I switched these to  target link options for OHL.

apple1417 commented on mingw-w64-x86_64.cmake:
```diff
@@ -0,0 +1,26 @@
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+
+#Useful later
+set(CMAKE_MINGW32 True)
```
There's a builtin var `MINGW`, so you don't need this.

abramhindle replied:
OK Good to know.

abramhindle replied:
I addressed this.

apple1417 commented on CMakeLists.txt:
```diff
@@ -12,14 +12,25 @@ option(BUILD_SHARED_LIBS "" OFF)
+if (DEFINED CMAKE_MINGW32)
+    set(CMAKE_CXX_EXTENSIONS False)
+    find_library(shlwapi libshlwapi REQUIRED)
+    set(CMAKE_EXE_COMPILER_FLAGS  "${CMAKE_EXE_COMPILER_FLAGS} -std=c++17 -pthread")
+    set(CMAKE_CXX_FLAGS "-std=c++17 ${CMAKE_CXX_FLAGS}")
+    set(CMAKE_CXX_STANDARD_LIBRARIES "-static-libgcc -static-libstdc++ -lwsock32 -lws2_32 ${CMAKE_CXX_STANDARD_LIBRARIES}")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-Bstatic,--whole-archive -lwinpthread -Wl,--no-whole-archive -pthread -lshlwapi")
```
Some of these flags are redundant (e.g. `-std=c++17` should be set by the `CMAKE_CXX_STANDARD` at the top), and the rest should be specified through `target_compile_options` or similar. This is something that can wait until we're sure what all's needed though, can compare what you and I need to get it working.

abramhindle replied:
there are issues with -std=gnu++17 and I was just trying to ensure the right flags were used.

abramhindle replied:
Addressed.

apple1417 commented on CMakeLists.txt:
```diff
@@ -29,7 +40,12 @@ file(GLOB_RECURSE sources CONFIGURE_DEPENDS "src/*.c" "src/*.cpp" "src/*.h" "src
 target_include_directories(OpenHotfixLoader PRIVATE "${PROJECT_BINARY_DIR}/inc" "src")
 
-target_link_libraries(OpenHotfixLoader PUBLIC minhook shlwapi cpr::cpr plog)
+if (DEFINED CMAKE_MINGW32)
+   target_link_libraries(OpenHotfixLoader PUBLIC minhook libshlwapi.a cpr::cpr plog)
+else()
+   target_link_libraries(OpenHotfixLoader PUBLIC minhook shlwapi cpr::cpr plog)
+endif()
```
Probably a bit cleaner to have one link at the top with the constant libs,, and then only shlwapi in the if

abramhindle replied:
I didn't make a list :/ I just made some vars. Why do libs have 2 names (lib name + link name).

abramhindle commented:
I use update-alternatives to enforce the use of -posix by default. It seems to me that the current setup ignores a lot of the cmake directives.

```
@slate:~/src/OpenHotfixLoader/out/mingw$ sudo update-alternatives --config x86_64-w64-mingw32-g++
There are 2 choices for the alternative x86_64-w64-mingw32-g++ (providing /usr/bin/x86_64-w64-mingw32-g++).

  Selection    Path                                   Priority   Status
------------------------------------------------------------
  0            /usr/bin/x86_64-w64-mingw32-g++-win32   60        auto mode
* 1            /usr/bin/x86_64-w64-mingw32-g++-posix   30        manual mode
  2            /usr/bin/x86_64-w64-mingw32-g++-win32   60        manual mode

Press <enter> to keep the current choice[*], or type selection number: 1

```

Here's a bit of the output you asked for:

```
[ 81%] Building CXX object CMakeFiles/OpenHotfixLoader.dir/src/loader.cpp.obj
/usr/bin/x86_64-w64-mingw32-g++ -DCURL_STATICLIB -DDEBUG -DOpenHotfixLoader_EXPORTS -DUNICODE -D_UNICODE @CMakeFiles/OpenHotfixLoader.dir/includes_CXX.rsp -std=c++17  -g -std=c++17 -Winvalid-pch -include /home/hindle1/src/OpenHotfixLoader/out/mingw/CMakeFiles/OpenHotfixLoader.dir/cmake_pch.hxx -MD -MT CMakeFiles/OpenHotfixLoader.dir/src/loader.cpp.obj -MF CMakeFiles/OpenHotfixLoader.dir/src/loader.cpp.obj.d -o CMakeFiles/OpenHotfixLoader.dir/src/loader.cpp.obj -c /home/hindle1/src/OpenHotfixLoader/src/loader.cpp
```

abramhindle committed:
Cleaned up mingw-w64-x86_64.cmake and CMakeLists.txt.
Reduced conditionals in the cmake file.
Removed some redundant flags.

abramhindle committed:
Move the mingw link options to target_link_libraries --- also make sure
pthreads are statically linked!

abramhindle commented:
So I moved the CXX options into a target specific link options. I have it compiling pthreads in there statically. I removed that extra mingw variable you didn't like. I tested it with Wonderlands COOP (Ubuntu 20.04, lutris, ge-7.1, Mingw, Epic Games). So take a peak. The final commit corrects link flags issues you had.

apple1417 committed:
few more mingw cleanups, add cmake preset

apple1417 commented:
Yup that got it compiling for me. Did a little more cleanup in 46ef41a2, including getting a preset working. I can compile fine with both msvc/mingw, if you could just check you can also still compile then I'll push it to your branch and merge (don't want to push something broken).

abramhindle commented:
I built mingw-debug and mingw-release successfully. And I booted up release and it did load hotfixes and my game didn't crash (lutris, ge-7.1, ubuntu 20.04, epic games, nvidia 470.141.03).

So looks good to me, if everything works for you then we are good to merge :) Thanks for the clean-up and feedback.

e: I pushed it my branch already.

```
Loading Plugin: C:\Program Files\Epic Games\TinyTinasWonderlands\OakGame\Binaries\Win64\Plugins\OpenHotfixLoader.dll


==> OpenHotfixLoader.log <==
2022-09-09 10:52:40.839 INFO  [1208] [ohl::hooks::init@358] [OHL] Hooks injected successfully
2022-09-09 10:53:46.684 INFO  [1272] [ohl::processing::handle_get_verification@222] [OHL] Starting to reload mods
2022-09-09 10:53:46.695 INFO  [2632] [ohl::loader::reload_impl@493] [OHL] Loaded files:
barfbunnies.wlhotfix,
chaos_gear_spawns_83_regular_15_primordial_2_ascended.wlhotfix,
limited-bank-and-backpack.wlhotfix,

2022-09-09 10:53:46.954 INFO  [1272] [ohl::processing::handle_discovery_from_json@276] [OHL] Injected hotfixes
2022-09-09 10:53:49.194 INFO  [1272] [ohl::processing::handle_news_from_json@375] [OHL] Injected news
2022-09-09 10:53:49.519 INFO  [1272] [ohl::processing::handle_add_image_to_cache@387] [OHL] Prevented news icon from being cached: https://raw.githubusercontent.com/apple1417/OpenHotfixLoader/master/news_icon.png

==> ../PluginLoader.log <==
Exiting Process...
```
